pkg/build/admission/secretinjector/admission.go

```diff
@@ -0,0 +1,108 @@
+package secretinjector
+
+import (
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/golang/glog"
+	authclient "github.com/openshift/origin/pkg/auth/client"
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	"github.com/openshift/origin/pkg/util/urlpattern"
+
+	"k8s.io/kubernetes/pkg/admission"
+	"k8s.io/kubernetes/pkg/api"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	coreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/unversioned"
+	"k8s.io/kubernetes/pkg/client/restclient"
+)
+
+func init() {
+	admission.RegisterPlugin("openshift.io/BuildConfigSecretInjector", func(c clientset.Interface, config io.Reader) (admission.Interface, error) {
+		return &secretInjector{
+			Handler: admission.NewHandler(admission.Create),
+		}, nil
+	})
+}
+
+type secretInjector struct {
+	*admission.Handler
+	restClientConfig restclient.Config
+}
+
+func (si *secretInjector) Admit(attr admission.Attributes) (err error) {
+	bc, ok := attr.GetObject().(*buildapi.BuildConfig)
+	if !ok {
+		return nil
+	}
+
+	if bc.Spec.Source.SourceSecret != nil || bc.Spec.Source.Git == nil {
+		return nil
+	}
+
+	// TODO: Jenkins bootstrapper verifies attr.GetSubresource() and attr.GetResource() - do I need to?
+
+	// TODO: no idea if the impersonation is necessary or correct?
```

Comment:
> yes i think we do want to impersonate to ensure we're only finding secrets the user can see. @deads2k would be the one to confirm this is being done correctly though.

Secret references aren't usually gated by user permissions.  See pods and secret volumes as an example.  They're optionally gated on service account powers.